runtime-modules/storage/src/lib.rs

```diff
@@ -3942,4 +3602,342 @@ impl<T: Trait> Module<T> {
+                Ok(correct)
+            }
+            BagOperationParams::<T>::Update(..) => tmp.collect(),
+            BagOperationParams::<T>::Delete => tmp
+                .map(|r| {
+                    r.map(|(id, mut bk)| {
+                        bk.unregister_bag_assignment();
+                        (id, bk)
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    fn generate_new_distribution_buckets(
+        bucket_ids: &BTreeSet<DistributionBucketId<T>>,
+        op: &BagOperationParams<T>,
+    ) -> BTreeMap<DistributionBucketId<T>, DistributionBucket<T>> {
+        bucket_ids
+            .iter()
+            .filter_map(|id| {
+                Self::ensure_distribution_bucket_exists(id)
+                    .map(|mut bk| match op {
+                        BagOperationParams::<T>::Create(..) => {
+                            bk.register_bag_assignment();
+                            (id.clone(), bk)
+                        }
+                        BagOperationParams::<T>::Delete => {
+                            bk.unregister_bag_assignment();
+                            (id.clone(), bk)
+                        }
+                        _ => (id.clone(), bk), // no op in case of objects upload/delete operation
+                    })
+                    .ok()
+            })
+            .collect()
+    }
+
+    fn compute_net_prize(
+        init_value: NetDeletionPrize<T>,
+        num_obj_to_create: usize,
+        num_obj_to_delete: usize,
+    ) -> NetDeletionPrize<T> {
+        let amnt = T::DataObjectDeletionPrize::get();
+        init_value
+            .add_balance(
+                iter::repeat(amnt)
+                    .take(num_obj_to_create)
+                    .fold(BalanceOf::<T>::zero(), |acc, x| acc.saturating_add(x)),
+            )
+            .sub_balance(
+                iter::repeat(amnt)
+                    .take(num_obj_to_delete)
+                    .fold(BalanceOf::<T>::zero(), |acc, x| acc.saturating_add(x)),
+            )
+    }
+
+    /// Utility function that checks existence for bag deletion / update &
+    /// verifies that dynamic bag does not exists in case of bag creation
+    /// returning a new one
+    fn retrieve_dynamic_bag(bag_op: &BagOperation<T>) -> Result<Bag<T>, DispatchError> {
+        if let BagOperationParams::<T>::Create(deletion_prize, _) = bag_op.params {
+            // map non existing dynamic bag to a new bag and existing bag into error
+            let dyn_bag_id = if let BagId::<T>::Dynamic(id) = &bag_op.bag_id {
+                Ok(id.clone())
+            } else {
+                Err(Error::<T>::CannotCreateStaticBag)
+            }?;
+
+            Self::ensure_bag_exists(&bag_op.bag_id).map_or_else(
+                |_| {
+                    Ok(Bag::<T>::default()
+                        .with_prize(deletion_prize)
+                        .with_storage_buckets(StorageBucketPicker::<T>::pick_storage_buckets(
```

Comment:
Why did we remove the bucket selection with the correct object size and number? 

`StorageBucketPicker::<T>::pick_storage_buckets(bag_type, voucher_update)`

Reply:
Pattern used by the `try_mutating_storage` : 
1. Get initial bag state: return bag on storage if present else create default one by selecting **some** storage bucket
2. Construct creation & removal parameters 
3. create `voucher_update` with params @ (2)
4. Try update storage buckets with `voucher_update` @ (3) by checking that storage bucket object number and total size bounds are respected. If it successful return a `BTreeMap` storage bucket id x storage bucket else throw.
Analogous with distribution bucket.
5. check sender/ treasury balance funds are sufficient for the bag operation specified
6. Update chain storage & balance after the mutation safe mark

This is the common pattern of all `*-bag` extrinsic. But:
1. Is complex and the resulting function is long and complex
2. It is definitely not obvious
So this makes me question having introduced such complex artifact

Reply:
:+1: